### PR TITLE
Adding support for case insensitive metadata keys on encrypted blobs

### PR DESCRIPTION
**Pull request by ibrahimrabab**
# Description

Adding support for the encryption metadata key **encryptiondata** to be case-insensitive, which is used for Blobs client-side encryption.

### CODE REVIEW

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 1]
```diff
@@ -733,11 +734,16 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
 
                 requestConditionsFinal.setIfMatch(response.getValue().getETag());
                 Mono<T> result = downloadCall.get();
-                if (response.getValue().getMetadata().get(ENCRYPTION_DATA_KEY) != null) {
+
+                Map<String, String> caseInsensitiveMetadata = new HashMap<>();
+                for (Map.Entry<String, String> entry : response.getValue().getMetadata().entrySet()) {
+                    caseInsensitiveMetadata.put(entry.getKey().toLowerCase(Locale.ROOT), entry.getValue());
+                }
+                if (caseInsensitiveMetadata.get(ENCRYPTION_DATA_KEY) != null) {
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:742]: I think wee also need to convert `ENCRYPTION_DATA_KEY`'s value to lower case here.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobClient.java [hunk 2]
```diff
@@ -486,9 +488,14 @@ private Context populateRequestConditionsAndContext(BlobRequestConditions reques
             this.getPropertiesWithResponse(requestConditions, timeout, context).getValue();
 
         requestConditions.setIfMatch(initialProperties.getETag());
-        if (initialProperties.getMetadata().get(ENCRYPTION_DATA_KEY) != null) {
+        Map<String, String> caseInsensitiveMetadata = new HashMap<>();
+        for (Map.Entry<String, String> entry : initialProperties.getMetadata().entrySet()) {
+            caseInsensitiveMetadata.put(entry.getKey().toLowerCase(Locale.ROOT), entry.getValue());
+        }
+
+        if (caseInsensitiveMetadata.get(ENCRYPTION_DATA_KEY) != null) {
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobClient.java:496]: Same here.

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java [hunk 3]
```diff
@@ -1596,6 +1598,58 @@ public void perCallPolicy() {
             client.getPropertiesWithResponse(null, null, null).getHeaders().getValue("x-ms-version"));
     }
 
+    @ParameterizedTest
+    @ValueSource(strings = {"ENCRYPTIONDATA", "EncryptionData", "eNcRyPtIoNdAtA"})
+    public void encryptionDataCaseInsensitivity(String newKey) {
+        byte[] data = getRandomByteArray(Constants.KB);
+        bec = getEncryptionClient(EncryptionVersion.V2, generateBlobName());
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java:1605]: This test looks really good, can we also test V1?  We may be able to do this by adding another parameter to this test and the test below.

**Comment by seanmcc-msft** [comment]
Also please update the changelogs.

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java [hunk 4]
```diff
@@ -1596,6 +1598,58 @@ public void perCallPolicy() {
...
+        // now call downloadStreamWithResponse with BlobRange passed to ensure ranged download is working
+        os = new ByteArrayOutputStream();
+        bec.downloadStreamWithResponse(os, new BlobRange(0, (long) Constants.KB), null, null, false, null,
+            Context.NONE);
+        assertArrayEquals(data, os.toByteArray());
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"ENCRYPTIONDATA", "EncryptionData", "eNcRyPtIoNdAtA"})
+    public void encryptionDataCaseInsensitivityAsyncClient(String newKey) {
```
Comment (seanmcc-msft) [inline sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java:1627]: Also please make sure to manually run these tests with Fiddler enabled to verify the encryption data metadata key has incorrect case sensitivity and we can still decrypt.

File: sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java (hunk 3 above)
Comment (ibrahimrabab) [reply sdk/storage/azure-storage-blob-cryptography/src/test/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlockBlobApiTests.java]: Added tests for V1 as well

File: sdk/storage/azure-storage-blob-cryptography/CHANGELOG.md [hunk 5]
```diff
@@ -7,6 +7,7 @@
 ### Breaking Changes
 
 ### Bugs Fixed
+- Fixed bug where a blob's contents could not be encrypted if the metadata's key set on the blob had any uppercase letters. The metadata key is now able to handle any casing.
```
Comment (jalauzon-msft) [inline sdk/storage/azure-storage-blob-cryptography/CHANGELOG.md:10]: nit: decrypted, not encrypted.

**Review by jalauzon-msft — APPROVED** [review]
Just one minor change to CHANGELOG, otherwise looks good.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 6]
```diff
@@ -733,11 +734,16 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
 
                 requestConditionsFinal.setIfMatch(response.getValue().getETag());
                 Mono<T> result = downloadCall.get();
-                if (response.getValue().getMetadata().get(ENCRYPTION_DATA_KEY) != null) {
+
+                Map<String, String> caseInsensitiveMetadata = new HashMap<>();
+                for (Map.Entry<String, String> entry : response.getValue().getMetadata().entrySet()) {
+                    caseInsensitiveMetadata.put(entry.getKey().toLowerCase(Locale.ROOT), entry.getValue());
+                }
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:741]: If the only requirement is to get `ENCRYPTION_DATA_KEY` in a case-insensitive way I would change this design a bit. We don't need to create the case insensitive map as it's just thrown away and could do something like:

```java
static String getEncryptionDataKey(Map<String, String> metadata) {
    if (CoreUtils.isNullOrEmpty(metadata) {
        return null;
    }

    for (Map.Entry<String, String> entry : metadata.entrySet()) {
        if (entry.getKey().length() != ENCRYPTION_DATA_KEY.length()) {
            continue;
        }

        if (ENCRYPTION_DATA_KEY.regionMatches(true, 0, entry.getKey(), 0, ENCRYPTION_DATA_KEY.length()) {
            return entry.getData();
        }
    }
}
```

This removes all the string lowercasing, map insertion, and then the final map get with a single pass over all entries in the map filtering out those which don't match ENCRYPTION_DATA_KEY's length and returns when the key matches case insensitively.

File: sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java [hunk 7]
```diff
@@ -733,17 +734,31 @@ private <T> Mono<T> populateRequestConditionsAndContext(BlobRequestConditions re
...
+                String encryptionDataKey = getEncryptionDataKey(response.getValue().getMetadata());
+                if (encryptionDataKey != null) {
+                    result = result.contextWrite(context -> context.put(ENCRYPTION_DATA_KEY,
+                        EncryptionData.getAndValidateEncryptionData(encryptionDataKey, requiresEncryption)));
                 }
                 return result;
             });
     }
 
+    private static String getEncryptionDataKey(Map<String, String> metadata) {
```
Comment (alzimmermsft) [inline sdk/storage/azure-storage-blob-cryptography/src/main/java/com/azure/storage/blob/specialized/cryptography/EncryptedBlobAsyncClient.java:747]: One more nit, this is repeated multiple times in the same package, let's make this a shared utility method so we don't have skew on the implementations over time.

**Review by alzimmermsft — APPROVED** [review]

**Comment by ibrahimrabab** [comment]
Overriding since failing tests are unrelated to the changes made.

**Pull request merged**